### PR TITLE
adds ip validation check in getK8sFreeIP

### PR DESCRIPTION
**Pull request by MarioBassem**
## Related issues:
- #394

### CODE REVIEW

**Review by xmonader — CHANGES_REQUESTED** [review]
if you want to convert to byte, do it inside of the loop not in the initializer and the condition, The check on the ip should happen very early in the function, before the slice append or maybe even outside of the isInByte check or the loop even. The whole function seems to be confusing ( I assume you're working against the last segment of an ip of /24 cidr? also if that's the case the check against the last segment, it should be called hostID )

File: internal/provider/resource_k8s.go [hunk 1]
```diff
@@ -989,6 +989,9 @@ func (k *K8sDeployer) getK8sFreeIP(ipRange gridtypes.IPNet, nodeID uint32) (stri
 		if !isInByte(k.NodeUsedIPs[nodeID], i) {
 			k.NodeUsedIPs[nodeID] = append(k.NodeUsedIPs[nodeID], i)
 			ip := ipRange.IP.To4()
+			if ip == nil {
```
Comment (xmonader) [inline internal/provider/resource_k8s.go:992]: if you want to convert to byte, do it inside of the loop not in the initializer and the condition, The check on the ip should happen very early in the function, before the slice append or maybe even outside of the isInByte check or the loop even. The whole function seems to be confusing ( I assume you're working against the last segment of an ip of /24 cidr? also if that's the case the check against the last segment, it should be called hostID )

**Review by xmonader — APPROVED** [review]

File: internal/provider/resource_k8s.go [hunk 2]
```diff
@@ -985,14 +985,15 @@ func (k *K8sNodeData) GenerateK8sWorkload(deployer *K8sDeployer, masterIP string
...
-			if ip == nil {
-				return "", fmt.Errorf("The provided ip range (%s) is not a valid ipv4.", ipRange.String())
-			}
-			ip[3] = i
+	ip := ipRange.IP.To4()
+	if ip == nil {
+		return "", fmt.Errorf("The provided ip range (%s) is not a valid ipv4.", ipRange.String())
+	}
+
+	for hostID := byte(2); hostID <= byte(255); hostID++ {
```
Comment (xmonader) [inline internal/provider/resource_k8s.go:993]: please keep the initializer and the loop conditions as ints, do the type conversion when needed in the `isInByte` call

**Review by xmonader — CHANGES_REQUESTED** [review]
added a minor comment

**Review by xmonader — APPROVED** [review]

File: internal/provider/resource_k8s.go [hunk 3]
```diff
@@ -985,11 +985,16 @@ func (k *K8sNodeData) GenerateK8sWorkload(deployer *K8sDeployer, masterIP string
...
-		if !isInByte(k.NodeUsedIPs[nodeID], i) {
-			k.NodeUsedIPs[nodeID] = append(k.NodeUsedIPs[nodeID], i)
-			ip := ipRange.IP.To4()
-			ip[3] = i
+	ip := ipRange.IP.To4()
+	if ip == nil {
+		return "", fmt.Errorf("The provided ip range (%s) is not a valid ipv4.", ipRange.String())
+	}
+
+	for i := 2; i <= 255; i++ {
```
Comment (AbdelrahmanElawady) [inline internal/provider/resource_k8s.go:993]: i think 255 shouldn't be included as it's reserved for broadcast

Comment (xmonader) [reply internal/provider/resource_k8s.go:993]: good catch!

**Pull request merged**